### ISSUE TITLE
Problems with Dune and ocamldebug

### ISSUE THREAD

**Issue by ramsdell** [ocaml/dune#3818]
This is most likely a problem with the Dune documentation, but it
could be something deeper.  I am the author of the
[Chase Model Finder](https://github.com/ramsdell/chase).
It uses Dune to build and install the package.  It works like a charm
for this task.

While developing the program, I found I had to resort to using
ocamlbuild.  I couldn't figure out how to get Dune to build a debugable
version of the executable along with a library of specially designed
pretty printers that display internal data structures in a human
understandable way.

It is hard for me to imagine that what I seek cannot be done now with
Dune, however, I still cannot figure out how from the documentation.
For ocamlbuild, the two ways of building the programs are specified in
the files, `chase.itarget` and `chase.d.itarget`.  Please add to the
Dune documentation an example that shows how one would create a
debugable version of a program along with a pretty printer library for
use with ocamldebug that parallels what I did with ocamlbuild.

**Comment by rgrinberg** [comment]
Is there any documentation for `d.itarget` in ocamlbuild? I'd like to see how they make debuggable versions of executables.

**Comment by ramsdell** [comment]
For Chase, the two files have been enclosed.  (I added the .txt extension to allow the files to be accepted.)

[chase.itarget.txt](https://github.com/ocaml/dune/files/5285452/chase.itarget.txt)

[chase.d.itarget.txt](https://github.com/ocaml/dune/files/5285454/chase.d.itarget.txt)

**Comment by ramsdell** [comment]
And here is the [ocamlbuild](https://nicolaspouillard.fr/ocamlbuild/ocamlbuild-user-guide.html) documentation.

**Comment by ejgallego** [comment]
In Coq we have some targets that work with `ocamldebug`, they are a bit "rusty" tho, see https://github.com/coq/coq/blob/master/dev/dune#L10

**Comment by rgrinberg** [comment]
Does ocamlbuild do something special to load pretty printers? I've looked at the [documentation](https://github.com/ocaml/ocamlbuild/blob/master/manual/manual.adoc#deprecated-targets) and it seems like it's just adding `-g`. Dune already does that and I don't see how `.d` is adding pretty printers anywhere. Perhaps @gasche knows how this works?

**Comment by gasche** [comment]
Indeed, what `foo.d.byte` does it just to build the program with the `-g` flag, and I believe that dune already does this by default. But then if you look at [chase.d.itarget](https://github.com/ramsdell/chase/blob/master/chase.d.itarget) in the OP's project, it is an aggregate target that build debug-enabled binaries *and* a `printers.cma` file, that contains pretty-printer for the project's data structures (see [printers.ml](https://github.com/ramsdell/chase/blob/master/printers.ml)). I am not an `ocamldebug` expert, but my understanding is that it can load this bytecode archive file, and then use that in directives to install more printers (see the instructions in comment in printers.ml; this is something that a user can feed manually to ocamldebug).

From a dune point of view, I think that (in addition to `-g`) what @ramsdell would want is an easy way to build the `.cma`, and then load it within `ocamldebug` (a first step, if that works, would be give ocamldebug the path within dune's artifact repositories). This probably not require any magic new feature, but some documentation / a simple example project would be helpful.

Emilio's Coq rule are an example of doing this: the [build rules](https://github.com/coq/coq/blob/master/dev/dune#L10) build the printer libraries, and then there is a [script](https://github.com/coq/coq/blob/master/dev/dune-dbg.in) to run ocamldebug with the right archives loaded (using ocamlfind, it seems).

**Comment by ramsdell** [comment]
Thank you Emilio for your example.  However I still don't get how to use it.  The goal is to add to the dune file that specifies how to build and install the product, some stanzas that construct a debuggable version of the one of the programs and a library that is to be loaded into ocamldebug.  None of these files are meant to be installed.

Here is what I tried to do.  I added the following stanza to the chase dune file.

```
(library
  (name printers)
  (public_name printers)
  (synopsis "Debug Printers")
  (wrapped false) 
  (modules Prelude Id Formula Structure Match Cong_close Normalize
	   Step Solve Fact Form_printer Fact_printer Printers))
```

Building with dune produces the following error message.

```sh
$ dune build
File "dune", line 24, characters 14-22:
24 |  (public_name printers)
                   ^^^^^^^^
Error: The current scope doesn't define package "printers".
The only packages for which you can declare elements to be installed in this
directory are:
- chase (because of chase.opam)
$
```

How do you build libraries that are not meant to be installed?

**Comment by ejgallego** [comment]
Hi @ramsdell , sorry for my late reply, actually you want to either fix your public name (should be of the form `pkg.lib`) or remove the public_name field entirely.

**Comment by Alizter** [comment]
@richardlford this one also seems relevant.
